src/electionguard/elgamal.py

```diff
@@ -154,6 +206,71 @@ def elgamal_encrypt(
     return ElGamalCiphertext(pad, data)
 
 
+_MAC_KEY_SIZE = 256
+
+
+def hashed_elgamal_encrypt(
+    message: bytes,
+    nonce: ElementModQ,
+    public_key: ElGamalPublicKey,
+    encryption_seed: ElementModQ,
+):
+    """
+    Encrypts a variable length byte message with a given random nonce and an ElGamal public key.
+
+    :param message: message (m) to encrypt; must be in bytes.
+    :param nonce: Randomly chosen nonce in [1, Q).
+    :param public_key: ElGamal public key.
+    :param encryption_seed: Encryption seed (Q) for election.
+    """
+
+    pad = g_pow_p(nonce)
+    pubkey_pow_n = pow_p(public_key, nonce)
+
+    session_key = hash_elems(pad, pubkey_pow_n)
+
+    (message_chunks, bit_length) = _get_chunks(message)
+    data = b""
+    for i, block in enumerate(message_chunks):
+        data_key = get_hmac(
+            session_key.to_hex_bytes(),
+            encryption_seed.to_hex_bytes(),
+            bit_length,
+            (i + 1),
+        )
+        data += bytes([a ^ b for (a, b) in zip(block, data_key)])
+
+    mac_key = get_hmac(
+        session_key.to_hex_bytes(), encryption_seed.to_hex_bytes(), _MAC_KEY_SIZE
+    )
+    to_mac = pad.to_hex_bytes() + data
+    mac = get_hmac(mac_key.to_hex_bytes(), to_mac.to_hex_bytes())
+
+    log_info(f": publicKey: {public_key.to_hex()}")
+    log_info(f": pad: {pad.to_hex()}")
+    log_info(f": data: {data}")
+    log_info(f": mac: {mac}")
+
+    return HashedElGamalCiphertext(pad, data, mac)
+
+
+_BLOCK_SIZE = 32
+
+
+def _get_chunks(message: bytes):
```

Comment:
This chunking function is to reduce some code complexity. It's not the prettiest for sure. Probably could use some work.

Reply:
moved.